dropshot/src/websocket.rs

```diff
@@ -0,0 +1,411 @@
+fn derive_accept_key(request_key: &[u8]) -> String {
+    // ... field is constructed by concatenating /key/ ...
+    // ... with the string "258EAFA5-E914-47DA-95CA-C5AB0DC85B11" (RFC 6455)
+    const WS_GUID: &[u8] = b"258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
+    let mut sha1 = Sha1::default();
+    sha1.update(request_key);
+    sha1.update(WS_GUID);
+    base64::encode(&sha1.finalize())
+}
+
+/**
+ * This `Extractor` implementation constructs an instance of `WebsocketUpgrade`
+ * from an HTTP request, and returns an error if the given request does not
+ * contain websocket upgrade headers.
+ */
+#[async_trait]
+impl Extractor for WebsocketUpgrade {
+    async fn from_request<Context: ServerContext>(
+        rqctx: Arc<RequestContext<Context>>,
+    ) -> Result<Self, HttpError> {
+        let request = &mut *rqctx.request.lock().await;
+
+        if !request
+            .headers()
+            .get(header::CONNECTION)
+            .and_then(|hv| hv.to_str().ok())
+            .map(|hv| {
+                hv.split(|c| c == ',' || c == ' ')
+                    .any(|vs| vs.eq_ignore_ascii_case("upgrade"))
+            })
+            .unwrap_or(false)
+        {
+            return Err(HttpError::for_bad_request(
+                None,
+                "expected connection upgrade".to_string(),
+            ));
+        }
+
+        if !request
+            .headers()
+            .get(header::UPGRADE)
+            .and_then(|v| v.to_str().ok())
+            .map(|v| {
+                v.split(|c| c == ',' || c == ' ')
+                    .any(|v| v.eq_ignore_ascii_case("websocket"))
+            })
+            .unwrap_or(false)
+        {
+            return Err(HttpError::for_bad_request(
+                None,
+                "unexpected protocol for upgrade".to_string(),
+            ));
+        }
+
+        if request
+            .headers()
+            .get(header::SEC_WEBSOCKET_VERSION)
+            .map(|v| v.as_bytes())
+            != Some(b"13")
+        {
+            return Err(HttpError::for_bad_request(
+                None,
+                "missing or invalid websocket version".to_string(),
+            ));
+        }
+
+        let accept_key = request
+            .headers()
+            .get(header::SEC_WEBSOCKET_KEY)
+            .map(|hv| hv.as_bytes())
+            .map(|key| derive_accept_key(key))
+            .ok_or_else(|| {
+                HttpError::for_bad_request(
+                    None,
+                    "missing websocket key".to_string(),
+                )
+            })?;
+
+        let method = request.method().as_str().to_string();
+        let route = request.uri().to_string();
+        let upgrade_fut = upgrade::on(request);
+        let ws_log = rqctx.log.new(o!(
+            "upgrade" => "websocket".to_string(),
+            "request_id" => rqctx.request_id.to_string(),
+            "method" => method,
+            "uri" => route.clone(),
```

Comment:
These should be already be inherited from `rqctx` I thought. +1 on the upgrade one tho, useful to differentiate which logs are from what component.

Reply:
gotcha -- also fwiw in case it's not clear, this one is just used by the `error!` calls in the wrapper around the user-provided future in `WebsocketUpgrade::handle`; if the user needs to log in their future, they can obtain it from rqctx the same way they do in any other endpoint & differentiate it however they wish

Reply:
ah, you're right. that makes sense!